### DISCUSSION TITLE
Missing npm dependency in serverless function (with yarn)

### DISCUSSION THREAD

**Discussion by elliottsj** [vercel/vercel#5846] [Help] [answered]
Certain npm dependencies are missing in my Vercel function deployment https://vercel.com/elliottsj/ackee.

Specifically, the `ts-tiny-invariant` package is missing and raises this error when a request comes in to `/api`:

```
{
  "errorType": "Error",
  "errorMessage": "Cannot find module 'ts-tiny-invariant'\nRequire stack:\n- /var/task/node_modules/ts-is-defined/dist/umd/index.js\n- /var/task/node_modules/@graphql-tools/mock/index.cjs.js\n- /var/task/node_modules/graphql-tools/index.cjs.js\n- /var/task/src/types/index.js\n- /var/task/src/utils/createApolloServer.js\n- /var/task/src/serverless.js\n- /var/task/api/index.js\n- /var/task/___vc_launcher.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js",
  "code": "MODULE_NOT_FOUND",
  "requireStack": [
    "/var/task/node_modules/ts-is-defined/dist/umd/index.js",
    "/var/task/node_modules/@graphql-tools/mock/index.cjs.js",
    "/var/task/node_modules/graphql-tools/index.cjs.js",
    "/var/task/src/types/index.js",
    "/var/task/src/utils/createApolloServer.js",
    "/var/task/src/serverless.js",
    "/var/task/api/index.js",
    "/var/task/___vc_launcher.js",
    "/var/runtime/UserFunction.js",
    "/var/runtime/index.js"
  ],
  "stack": [
    "Error: Cannot find module 'ts-tiny-invariant'",
    "Require stack:",
    "- /var/task/node_modules/ts-is-defined/dist/umd/index.js",
    "- /var/task/node_modules/@graphql-tools/mock/index.cjs.js",
    "- /var/task/node_modules/graphql-tools/index.cjs.js",
    "- /var/task/src/types/index.js",
    "- /var/task/src/utils/createApolloServer.js",
    "- /var/task/src/serverless.js",
    "- /var/task/api/index.js",
    "- /var/task/___vc_launcher.js",
    "- /var/runtime/UserFunction.js",
    "- /var/runtime/index.js",
    "    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)",
    "    at Function.Module._load (internal/modules/cjs/loader.js:725:27)",
    "    at Module.require (internal/modules/cjs/loader.js:952:19)",
    "    at require (internal/modules/cjs/helpers.js:88:18)",
    "    at /var/task/node_modules/ts-is-defined/dist/umd/index.js:16:47",
    "    at /var/task/node_modules/ts-is-defined/dist/umd/index.js:6:17",
    "    at Object.<anonymous> (/var/task/node_modules/ts-is-defined/dist/umd/index.js:12:3)",
    "    at Module._compile (internal/modules/cjs/loader.js:1063:30)",
    "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)",
    "    at Module.load (internal/modules/cjs/loader.js:928:32)",
    "    at Function.Module._load (internal/modules/cjs/loader.js:769:14)",
    "    at Module.require (internal/modules/cjs/loader.js:952:19)",
    "    at require (internal/modules/cjs/helpers.js:88:18)",
    "    at Object.<anonymous> (/var/task/node_modules/@graphql-tools/mock/index.cjs.js:8:21)",
    "    at Module._compile (internal/modules/cjs/loader.js:1063:30)",
    "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)"
  ]
}
```

It seems that Vercel automatically omits unused dependencies from the file system for deployed functions, but `ts-tiny-invariant` is unexpectedly omitted in this case. To verify this, I've tried [adding a `require('ts-tiny-invariant')` statement to the function's entry point](https://github.com/elliottsj/Ackee/commit/4bbf9797b2fefa452c4418595837793f16a6a3e5) and the package becomes present again, and the error no longer occurs.

From the stack trace, we have the transitive dependency chain `src/types/index.js` -> `graphql-tools` -> `@graphql-tools/mock` -> `ts-is-defined` -> `ts-tiny-invariant`. The first four npm packages seem to be found correctly; only `ts-tiny-invariant` is missing.

I have also tried [adding a `require('graphql-tools')` statement to the entry point](https://github.com/elliottsj/Ackee/commit/031117b454ff3b8375bc16455a989c9e81b87aa0) (in case the behaviour has something to do with the `src/types/index.js` module), but the original error still occurs.

Help is appreciated 😄 

Context:
* https://github.com/electerious/Ackee/issues/223#issuecomment-778789424

**Comment by heysailor** [comment]
I'm curious, are you using Node 14 on Vercel, or Node 12?

... a similar error occurred on Vercel in production with Node 12, and was quickly fixed [following a discussion here](https://github.com/vercel/vercel/discussions/5132) - this involved erroneously removed dependencies referenced in child code modules.

**Reply by electerious** [reply]
Should be Node 12 as far is I know. Might be worth to give Node 14 a try, too.

**Reply by elliottsj** [reply]
I changed my build command to `node -v && yarn build`, which is printing `v14.15.4`.

The [engines config](https://github.com/electerious/Ackee/blob/01037bea6280995885c98b1833944511573d8799/package.json#L122) is `"node": ">= 12"`, so v14.15.4 is compatible

**Reply by heysailor** [reply]
@paulogdm you were involved last time, are you able to help here? Thanks.

It seems to be very similar to our discussion about missing dependencies when using yarn workspaces: https://github.com/vercel/vercel/discussions/5132

**Comment by heysailor** [comment, marked as answer]
@paulogdm https://github.com/vercel/vercel/discussions/5846#discussioncomment-375165 could you help?

**Reply by paulogdm** [reply]
Apologies. Our support team can handle that for sure. Just send them a repro on https://vercel.com/support/request

**Reply by elliottsj** [reply]
I checked again, and things seem to work now, even after removing the `require('ts-tiny-invariant')` statement. Seems like the bug has been fixed 🎉

**Reply by heysailor** [reply]
Excellent. You can mark this as answered now.

**Reply by fordneild** [reply]
@paulogdm
@heysailor 
Im still having this issue with the same ts-tiny-invariant package  using npm. 
Here is the currently deployed app with this issue: https://hyroji-b4sd00xjq-hyroji.vercel.app/
and the relevant logs
```(internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/var/task/node_modules/ts-is-defined/dist/umd/index.js',
    '/var/task/node_modules/@graphql-tools/mock/index.cjs.js',
    '/var/task/node_modules/neo4j-graphql-js/node_modules/graphql-tools/index.cjs.js',
    '/var/task/node_modules/neo4j-graphql-js/dist/augment/augment.js',
    '/var/task/node_modules/neo4j-graphql-js/dist/augment/fields.js',
    '/var/task/node_modules/neo4j-graphql-js/dist/utils.js',
    '/var/task/node_modules/neo4j-graphql-js/dist/translate/translate.js',
    '/var/task/node_modules/neo4j-graphql-js/dist/index.js',
    '/var/task/.next/serverless/pages/api/graphql.js',
    '/var/task/now__launcher.js',
    '/var/runtime/UserFunction.js',
    '/var/runtime/index.js'
  ]
}
```
and my package.json (where I even manually added ts-tiny-invariant to try and stop it from being remove)
```
{
  "name": "nan",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@apollo/client": "^3.3.9",
    "@apollo/react-hooks": "^4.0.0",
    "apollo-server-micro": "^2.20.0",
    "apollo-server-plugin-http-headers": "^0.1.4",
    "bcrypt": "^5.0.0",
    "graphql": "^15.5.0",
    "jsonwebtoken": "^8.5.1",
    "neo4j-driver": "^4.2.2",
    "neo4j-graphql-js": "^2.19.2",
    "next": "10.0.6",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "sass": "^1.32.7",
    "ts-tiny-invariant": "0.0.3"
  },
  "devDependencies": {
    "@types/bcrypt": "^3.0.0",
    "@types/cookie": "^0.4.0",
    "@types/jsonwebtoken": "^8.5.0",
    "@types/node": "^14.14.25",
    "@types/react": "^17.0.1",
    "typescript": "^4.1.5"
  }
}
```

**Comment by ramiel** [comment]
For me too, this is not fixed

**Comment by electerious** [comment]
This should be fixed now. At least in some cases.

Related Twitter discussion: https://twitter.com/rauchg/status/1401594243081146368

**Reply by ramiel** [reply]
I'm in the same situation of the twitter user, with an api not served through next and the fix, adding `FORCE_RUNTIME_TAG=canary` doesn't solve the issue for me, as you can see here https://beeth-core-nh9knbkv3-ramiel.vercel.app/

**Reply by perminder-klair** [reply]
still same issue. adding FORCE_RUNTIME_TAG=canary does not fix it

**Reply by MonoCredit** [reply]
Same issue !
Adding  FORCE_RUNTIME_TAG=canary didn't fix it.
Adding import 'ts-tiny-invariant' fixed it

**Reply by ramiel** [reply]
~~This isn't fixing the issue for me. Adding the import and/or the package itself to the list of dependencies makes no difference~~ (I was adding in the wrong place)

This is actually fixing the issue!

**Reply by chrisui** [reply]
Just to add a bit of noise as this is a highly critical issue, this is not working for us either. `FORCE_RUNTIME_TAG=canary` unfortunately makes no difference.
